File: src/personal_data_warehouse_photos/sync.py
"""Apple Photos upload runner.

Snapshot the library DB, resolve original-resource candidates, export each
selected resource through PhotoKit with iCloud network access enabled, and
upload its complete bytes through the app's shared photo endpoints (blob +
envelope). File blobs use resumable Drive sessions, so full originals are not
bounded by the app/proxy request ceiling. The local ``originals/`` cache is
never used as a source of truth.
"""

from __future__ import annotations

import hashlib
import tempfile
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path

from personal_data_warehouse_photos.envelope import build_photo_metadata, provenance_dedup_sha256
from personal_data_warehouse_photos.exporter import ExportedPhotoFile, PhotoKitAssetExporter
from personal_data_warehouse_photos.scanner import (
    PhotoFileCandidate,
    scan_photo_file_candidates,
    snapshot_photos_store,
)
from personal_data_warehouse_photos.state import SOURCE_TYPE_ASSET_FILE, PhotosUploadState

PHOTO_SOURCE = "apple_photos"

# Some rows in Photos.sqlite can never be exported through PhotoKit (burst
# stack members, records whose iCloud original is gone, ...). Retrying them
# every 30 minutes forever burned the run's export budget and, because the
# first failure was re-raised, kept every scheduled run red — so a handful of
# dead assets hid the health of the ~12k that upload fine. Failing files are
# therefore backed off exponentially, and after MAX_FATAL_ATTEMPTS they stop
# failing the run: they are reported loudly in the summary instead.
MAX_FATAL_ATTEMPTS = 5
RETRY_BACKOFF_BASE = timedelta(minutes=30)
MAX_RETRY_BACKOFF = timedelta(days=7)
# 2 ** 12 * 30min is far past the cap; bound the shift so the delay math cannot
# overflow for a file that has been failing for years.
MAX_BACKOFF_DOUBLINGS = 12


def retry_delay(failure_count: int) -> timedelta:
    """Exponential backoff for a file that has failed ``failure_count`` times."""
    if failure_count <= 0:
        return timedelta(0)
    doublings = min(failure_count, MAX_BACKOFF_DOUBLINGS) - 1
    return min(RETRY_BACKOFF_BASE * (2**doublings), MAX_RETRY_BACKOFF)


@dataclass(frozen=True)
class PhotosUploadSummary:
    assets_seen: int
    files_seen: int
    files_selected: int
    files_skipped: int
    files_exported: int
    files_uploaded: int
    metadata_uploaded: int
    bytes_exported: int = 0
    bytes_uploaded: int = 0
    files_deferred: int = 0
    files_failed: int = 0


class PhotosUploadRunner:
    def __init__(
        self,
        *,
        account: str,
        library_path: Path | str,
        ingest_client,
        logger,
        now=None,
        limit: int | None = None,
        mode: str = "incremental",
        upload_state: PhotosUploadState | None = None,
        before_upload_check=None,
        resource_exporter=None,
    ) -> None:
        if ingest_client is None:
            raise ValueError("ingest_client is required")
        if mode not in {"full", "incremental"}:
            raise ValueError("mode must be 'full' or 'incremental'")
        self._account = account
        self._library_path = Path(library_path).expanduser()
        self._ingest_client = ingest_client
        self._logger = logger
        self._now = now or (lambda: datetime.now(tz=UTC))
        self._limit = limit
        self._mode = mode
        self._upload_state = upload_state
        self._before_upload_check = before_upload_check
        self._resource_exporter = resource_exporter or PhotoKitAssetExporter()

    def sync(self) -> PhotosUploadSummary:
        self._logger.info("Scanning Apple Photos library at %s", self._library_path)
        with tempfile.TemporaryDirectory(prefix="pdw-photos-") as working_dir:
            snapshot = snapshot_photos_store(self._library_path, working_dir)
            candidates = scan_photo_file_candidates(snapshot)
            return self._sync_candidates(candidates, export_dir=Path(working_dir) / "exports")

    def _sync_candidates(
        self,
        candidates: list[PhotoFileCandidate],
        *,
        export_dir: Path,
    ) -> PhotosUploadSummary:
        export_dir.mkdir(parents=True, exist_ok=True)
        assets_seen = len({candidate.native_id for candidate in candidates})
        self._logger.info(
            "Apple Photos inventory: assets=%s original_resources=%s (full bytes exported via PhotoKit)",
            assets_seen,
            len(candidates),
        )
        # v1 uploads originals only; edited renditions (Photos' adjusted
        # output under resources/renders) are a known follow-up. Keep the
        # count visible so the gap never reads as complete coverage.
        edited_assets = len(
            {
                candidate.native_id
                for candidate in candidates
                if candidate.role == "original" and candidate.apple_record.get("adjustments_state")
            }
        )
        if edited_assets:
            self._logger.info(
                "%s asset(s) have Photos adjustments; edited renditions are not uploaded yet (originals only)",
                edited_assets,
            )

        selected: list[PhotoFileCandidate] = []
        state_skipped = 0
        deferred: list[tuple[PhotoFileCandidate, datetime]] = []
        for candidate in candidates:
            if self._mode == "incremental" and self._is_state_complete(candidate):
                state_skipped += 1
                continue
            retry_at = self._retry_not_before(candidate)
            if retry_at is not None:
                deferred.append((candidate, retry_at))
                continue
            selected.append(candidate)
        # The limit applies AFTER state selection (unlike voice memos) so a
        # capped run always makes forward progress through the backlog instead
        # of re-considering the same already-complete head of the list.
        # Backed-off failures are dropped before the limit too, so broken files
        # never consume slots that working ones need.
        if self._limit is not None:
            selected = selected[: self._limit]

        self._logger.info(
            "Incremental selection: selected=%s skipped=%s deferred=%s",
            len(selected),
            state_skipped,
            len(deferred),
        )
        if deferred:
            self._log_deferred(deferred)
        if selected and self._before_upload_check is not None:
            skip_reason = self._before_upload_check()
            if skip_reason:
                self._logger.warning("Skipping photo upload: %s", skip_reason)
                return PhotosUploadSummary(
                    assets_seen=assets_seen,
                    files_seen=len(candidates),
                    files_selected=len(selected),
                    files_skipped=state_skipped,
                    files_exported=0,
                    files_uploaded=0,
                    metadata_uploaded=0,
                    files_deferred=len(deferred),
                )

        # Per-file failures are collected, not raised mid-batch (voice-memos
        # pattern): successes are recorded in upload_state so the next run
        # resumes past them, and a still-retryable failure re-raises at the end
        # so the run exits non-zero for the status helper.
        failures: list[tuple[PhotoFileCandidate, Exception]] = []
        uploaded = 0
        metadata_uploaded = 0
        exported_count = 0
        bytes_exported = 0
        bytes_uploaded = 0
        for index, candidate in enumerate(selected, start=1):
            exported: ExportedPhotoFile | None = None
            try:
                self._logger.info(
                    "[%s/%s] Exporting full original %s (%s) through PhotoKit",
                    index,
                    len(selected),
                    candidate.filename,
                    candidate.role,
                )
                exported = self._resource_exporter.export(candidate, export_dir)
                exported_count += 1
                bytes_exported += exported.size_bytes
                self._upload_candidate(
                    index=index,
                    total=len(selected),
                    candidate=candidate,
                    exported=exported,
                )
            except Exception as exc:  # noqa: BLE001 - surfaced after the batch
                self._logger.warning("Failed to upload %s: %s", candidate.filename, exc)
                failures.append((candidate, exc))
                continue
            finally:
                if exported is not None:
                    try:
                        exported.path.unlink(missing_ok=True)
                    except OSError as exc:
                        self._logger.warning(
                            "Could not remove temporary PhotoKit export %s: %s",
                            exported.path,
                            exc,
                        )
            uploaded += 1
            metadata_uploaded += 1
            bytes_uploaded += exported.size_bytes

        retryable = self._record_failures(failures)

        summary = PhotosUploadSummary(
            assets_seen=assets_seen,
            files_seen=len(candidates),
            files_selected=len(selected),
            files_skipped=state_skipped,
            files_exported=exported_count,
            files_uploaded=uploaded,
            metadata_uploaded=metadata_uploaded,
            bytes_exported=bytes_exported,
            bytes_uploaded=bytes_uploaded,
            files_deferred=len(deferred),
            files_failed=len(failures),
        )
        self._logger.info(
            "Photo upload summary: assets=%s original_resources=%s selected=%s exported=%s (%s) "
            "uploaded=%s (%s) skipped=%s deferred=%s failed=%s",
            summary.assets_seen,
            summary.files_seen,
            summary.files_selected,
            summary.files_exported,
            format_bytes(summary.bytes_exported),
            summary.files_uploaded,
            format_bytes(summary.bytes_uploaded),
            summary.files_skipped,
            summary.files_deferred,
            summary.files_failed,
        )
        if retryable:
            self._logger.warning(
                "Photo upload finished with %s failed file(s) after uploading %s; re-raising the "
                "first so the run is marked failed (successful uploads are recorded, so the next "
                "run resumes past them)",
                len(failures),
                summary.files_uploaded,
            )
            raise retryable[0]
        if failures:
            self._logger.warning(
                "%s file(s) have now failed %s+ times and no longer fail the run; they retry on "
                "backoff (up to %s days) and stay visible as failed= in this summary: %s",
                len(failures),
                MAX_FATAL_ATTEMPTS,
                MAX_RETRY_BACKOFF.days,
                ", ".join(f"{candidate.filename} ({exc})" for candidate, exc in failures[:5]),
            )
        return summary

    def _record_failures(
        self, failures: list[tuple[PhotoFileCandidate, Exception]]
    ) -> list[Exception]:
        """Persist each failed attempt; return the ones that must fail the run.

        A failure keeps failing the run until the same file has been attempted
        MAX_FATAL_ATTEMPTS times AND an upload has succeeded since that streak
        began. The second condition is what keeps a real outage (revoked Photos
        access, dead network) loudly red instead of quietly "green with
        failures": if nothing has succeeded since the failures started, every
        failure is still treated as fatal no matter how many attempts it has.
        """
        if self._upload_state is None:
            return [exc for _, exc in failures]
        now = self._now()
        latest_success = self._upload_state.latest_success_at()
        retryable: list[Exception] = []
        for candidate, exc in failures:
            entry = self._upload_state.mark_failure(
                source_type=SOURCE_TYPE_ASSET_FILE,
                source_id=candidate.state_id,
                fingerprint=candidate.fingerprint,
                error=str(exc),
                now=now,
            )
            first_failure = _parse_timestamp(entry.first_failure_at)
            proven = (
                latest_success is not None
                and first_failure is not None
                and latest_success >= first_failure
            )
            if entry.failure_count < MAX_FATAL_ATTEMPTS or not proven:
                retryable.append(exc)
        return retryable

    def _is_state_complete(self, candidate: PhotoFileCandidate) -> bool:
        if self._upload_state is None:
            return False
        return self._upload_state.is_complete(
            source_type=SOURCE_TYPE_ASSET_FILE,
            source_id=candidate.state_id,
            fingerprint=candidate.fingerprint,
        )

    def _retry_not_before(self, candidate: PhotoFileCandidate) -> datetime | None:
        """The moment a previously failed candidate may be attempted again."""
        if self._upload_state is None:
            return None
        entry = self._upload_state.entry_for(
            source_type=SOURCE_TYPE_ASSET_FILE, source_id=candidate.state_id
        )
        if entry is None or entry.failure_count <= 0:
            return None
        if entry.fingerprint != candidate.fingerprint:
            # The asset changed in Photos: a different file to upload, so the
            # old streak says nothing about it.
            return None
        last_failure = _parse_timestamp(entry.last_failure_at)
        if last_failure is None:
            return None
        retry_at = last_failure + retry_delay(entry.failure_count)
        return retry_at if self._now() < retry_at else None

    def _log_deferred(self, deferred: list[tuple[PhotoFileCandidate, datetime]]) -> None:
        earliest = min(retry_at for _, retry_at in deferred)
        self._logger.warning(
            "Deferred %s previously failed file(s) still in retry backoff (earliest retry %s): %s",
            len(deferred),
            earliest.isoformat(),
            ", ".join(candidate.filename for candidate, _ in deferred[:5]),
        )

    def _upload_candidate(
        self,
        *,
        index: int,
        total: int,
        candidate: PhotoFileCandidate,
        exported: ExportedPhotoFile,
    ) -> None:
        content_sha256 = sha256_file(exported.path)
        captured_at = candidate.captured_at or self._now().strftime("%Y-%m-%dT%H:%M:%S")
        self._logger.info(
            "[%s/%s] Uploading %s (%s, %s)",
            index,
            total,
            exported.filename,
            candidate.role,
            format_bytes(exported.size_bytes),
        )
        stored = self._ingest_client.upload_photo_file_path(
            exported.path,
            captured_at=captured_at,
            extension=exported.extension,
            content_type=exported.mime_type,
            content_sha256=content_sha256,
        )
        envelope = build_photo_metadata(
            source=PHOTO_SOURCE,
            account=self._account,
            native_id=candidate.native_id,
            role=candidate.role,
            filename=exported.filename,
            mime_type=exported.mime_type,
            size_bytes=exported.size_bytes,
            content_sha256=content_sha256,
            uploaded_at=self._now().isoformat(),
            width=candidate.width,
            height=candidate.height,
            captured_at=captured_at,
            capture_tz_offset=candidate.capture_tz_offset,
            camera_make=candidate.camera_make,
            camera_model=candidate.camera_model,
            record_key="apple_record",
            record=candidate.apple_record,
        )
        self._ingest_client.upload_photo_metadata(
            envelope,
            captured_at=captured_at,
            file_content_sha256=content_sha256,
            metadata_dedup_sha256=provenance_dedup_sha256(
                source=PHOTO_SOURCE,
                account=self._account,
                native_id=candidate.native_id,
                role=candidate.role,
                file_content_sha256=content_sha256,
            ),
        )
        if self._upload_state is not None:
            self._upload_state.mark_success(
                source_type=SOURCE_TYPE_ASSET_FILE,
                source_id=candidate.state_id,
                fingerprint=candidate.fingerprint,
                now=self._now(),
                content_sha256=content_sha256,
                storage_key=str(stored.get("storage_key", "")) if isinstance(stored, dict) else "",
            )


def _parse_timestamp(value: str) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value)
    except ValueError:
        return None


def format_bytes(count: float) -> str:
    size = float(count)
    for unit in ("B", "KiB", "MiB", "GiB", "TiB"):
        if size < 1024 or unit == "TiB":
            return f"{size:.1f} {unit}" if unit != "B" else f"{int(size)} B"
        size /= 1024
    return f"{size:.1f} TiB"


def sha256_file(path: Path, *, chunk_size: int = 8 * 1024 * 1024) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(chunk_size), b""):
            digest.update(chunk)
    return digest.hexdigest()

File: app/internal/uploaders/photos/sync.go
package photos

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/ingestclient"
	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

// PhotoSource is the envelope `source` slug the Dagster reader routes on.
const PhotoSource = "apple_photos"

// Some rows in Photos.sqlite can never be exported through PhotoKit (burst
// stack members, records whose iCloud original is gone, ...). Retrying them
// every 30 minutes forever burned the run's export budget and, because the
// first failure was re-raised, kept every scheduled run red, so a handful of
// dead assets hid the health of the ~12k that upload fine. Failing files are
// therefore backed off exponentially, and after MaxFatalAttempts they stop
// failing the run: they are reported loudly in the summary instead.
const (
	MaxFatalAttempts = 5
	RetryBackoffBase = 30 * time.Minute
	MaxRetryBackoff  = 7 * 24 * time.Hour
	// 2 ** 12 * 30min is far past the cap; bound the shift so the delay math
	// cannot overflow for a file that has been failing for years.
	MaxBackoffDoublings = 12
)

// RetryDelay is the exponential backoff for a file that has failed
// failureCount times.
func RetryDelay(failureCount int64) time.Duration {
	if failureCount <= 0 {
		return 0
	}
	doublings := failureCount
	if doublings > MaxBackoffDoublings {
		doublings = MaxBackoffDoublings
	}
	delay := RetryBackoffBase * (1 << (doublings - 1))
	if delay > MaxRetryBackoff {
		return MaxRetryBackoff
	}
	return delay
}

// Summary is the run's counts.
type Summary struct {
	AssetsSeen       int
	FilesSeen        int
	FilesSelected    int
	FilesSkipped     int
	FilesExported    int
	FilesUploaded    int
	MetadataUploaded int
	BytesExported    int64
	BytesUploaded    int64
	FilesDeferred    int
	FilesFailed      int
}

// Uploader is what the runner needs from the ingest client.
type Uploader interface {
	UploadPhotoFile(path, capturedAt, extension, contentType, contentSHA256 string) (ingestclient.StoredObject, error)
	UploadPhotoMetadata(payload map[string]any, capturedAt, fileContentSHA256, metadataDedupSHA256 string) (ingestclient.StoredObject, error)
}

// Runner snapshots the library DB, resolves original-resource candidates,
// exports each selected resource through PhotoKit with iCloud network access
// enabled, and uploads its complete bytes through the app's shared photo
// endpoints (blob + envelope). The local originals/ cache is never a source
// of truth.
type Runner struct {
	Account           string
	LibraryPath       string
	Client            Uploader
	Logger            common.Logger
	Now               func() time.Time
	Limit             int // 0 means no limit
	Mode              string
	State             *State
	BeforeUploadCheck func() string
	Exporter          Exporter
}

type failedCandidate struct {
	candidate Candidate
	err       error
}

type deferredCandidate struct {
	candidate Candidate
	retryAt   time.Time
}

// Sync runs one pass. A still-retryable per-file failure is returned after
// the whole batch so the run exits non-zero for the status helper, while
// every success stays recorded in the state file.
func (r *Runner) Sync() (Summary, error) {
	if r.Client == nil {
		return Summary{}, errors.New("ingest_client is required")
	}
	if r.Mode == "" {
		r.Mode = "incremental"
	}
	if r.Mode != "incremental" && r.Mode != "full" {
		return Summary{}, errors.New("mode must be 'full' or 'incremental'")
	}
	if r.Now == nil {
		r.Now = func() time.Time { return time.Now().UTC() }
	}
	if r.Exporter == nil {
		r.Exporter = &PhotoKitExporter{}
	}
	libraryPath := common.ExpandUser(r.LibraryPath)
	r.Logger.Infof("Scanning Apple Photos library at %s", libraryPath)
	workingDir, err := os.MkdirTemp("", "pdw-photos-")
	if err != nil {
		return Summary{}, err
	}
	defer os.RemoveAll(workingDir)
	snapshot, err := SnapshotStore(libraryPath, workingDir)
	if err != nil {
		return Summary{}, err
	}
	candidates, err := Scan(snapshot)
	if err != nil {
		return Summary{}, err
	}
	return r.syncCandidates(candidates, filepath.Join(workingDir, "exports"))
}

func (r *Runner) syncCandidates(candidates []Candidate, exportDir string) (Summary, error) {
	if err := os.MkdirAll(exportDir, 0o755); err != nil {
		return Summary{}, err
	}
	assets := map[string]bool{}
	edited := map[string]bool{}
	for _, candidate := range candidates {
		assets[candidate.NativeID] = true
		if candidate.Role == "original" && common.ToInt(candidate.AppleRecord["adjustments_state"]) != 0 {
			edited[candidate.NativeID] = true
		}
	}
	assetsSeen := len(assets)
	r.Logger.Infof("Apple Photos inventory: assets=%d original_resources=%d (full bytes exported via PhotoKit)", assetsSeen, len(candidates))
	// v1 uploads originals only; edited renditions (Photos' adjusted output
	// under resources/renders) are a known follow-up. Keep the count visible
	// so the gap never reads as complete coverage.
	if len(edited) > 0 {
		r.Logger.Infof("%d asset(s) have Photos adjustments; edited renditions are not uploaded yet (originals only)", len(edited))
	}

	var selected []Candidate
	var deferred []deferredCandidate
	stateSkipped := 0
	for _, candidate := range candidates {
		if r.Mode == "incremental" {
			complete, err := r.isStateComplete(candidate)
			if err != nil {
				return Summary{}, err
			}
			if complete {
				stateSkipped++
				continue
			}
		}
		retryAt, deferredNow, err := r.retryNotBefore(candidate)
		if err != nil {
			return Summary{}, err
		}
		if deferredNow {
			deferred = append(deferred, deferredCandidate{candidate, retryAt})
			continue
		}
		selected = append(selected, candidate)
	}
	// The limit applies AFTER state selection (unlike voice memos) so a
	// capped run always makes forward progress through the backlog instead
	// of re-considering the same already-complete head of the list.
	// Backed-off failures are dropped before the limit too, so broken files
	// never consume slots that working ones need.
	if r.Limit > 0 && len(selected) > r.Limit {
		selected = selected[:r.Limit]
	}
	r.Logger.Infof("Incremental selection: selected=%d skipped=%d deferred=%d", len(selected), stateSkipped, len(deferred))
	if len(deferred) > 0 {
		r.logDeferred(deferred)
	}
	summary := Summary{
		AssetsSeen:    assetsSeen,
		FilesSeen:     len(candidates),
		FilesSelected: len(selected),
		FilesSkipped:  stateSkipped,
		FilesDeferred: len(deferred),
	}
	if len(selected) > 0 && r.BeforeUploadCheck != nil {
		if reason := r.BeforeUploadCheck(); reason != "" {
			r.Logger.Warningf("Skipping photo upload: %s", reason)
			return summary, nil
		}
	}

	// Per-file failures are collected, not raised mid-batch (voice-memos
	// pattern): successes are recorded in upload_state so the next run
	// resumes past them, and a still-retryable failure re-raises at the end
	// so the run exits non-zero for the status helper.
	var failures []failedCandidate
	for index, candidate := range selected {
		r.Logger.Infof("[%d/%d] Exporting full original %s (%s) through PhotoKit", index+1, len(selected), candidate.Filename, candidate.Role)
		exported, err := r.Exporter.Export(candidate, exportDir)
		if err != nil {
			r.Logger.Warningf("Failed to upload %s: %s", candidate.Filename, err)
			failures = append(failures, failedCandidate{candidate, err})
			continue
		}
		summary.FilesExported++
		summary.BytesExported += exported.SizeBytes
		err = r.uploadCandidate(index+1, len(selected), candidate, exported)
		if removeErr := os.Remove(exported.Path); removeErr != nil && !os.IsNotExist(removeErr) {
			r.Logger.Warningf("Could not remove temporary PhotoKit export %s: %s", exported.Path, removeErr)
		}
		if err != nil {
			r.Logger.Warningf("Failed to upload %s: %s", candidate.Filename, err)
			failures = append(failures, failedCandidate{candidate, err})
			continue
		}
		summary.FilesUploaded++
		summary.MetadataUploaded++
		summary.BytesUploaded += exported.SizeBytes
	}

	retryable, err := r.recordFailures(failures)
	if err != nil {
		return summary, err
	}
	summary.FilesFailed = len(failures)
	r.Logger.Infof("Photo upload summary: assets=%d original_resources=%d selected=%d exported=%d (%s) uploaded=%d (%s) skipped=%d deferred=%d failed=%d",
		summary.AssetsSeen, summary.FilesSeen, summary.FilesSelected, summary.FilesExported, formatBytes(summary.BytesExported),
		summary.FilesUploaded, formatBytes(summary.BytesUploaded), summary.FilesSkipped, summary.FilesDeferred, summary.FilesFailed)
	if len(retryable) > 0 {
		r.Logger.Warningf("Photo upload finished with %d failed file(s) after uploading %d; re-raising the first so the run is marked failed (successful uploads are recorded, so the next run resumes past them)",
			len(failures), summary.FilesUploaded)
		return summary, retryable[0]
	}
	if len(failures) > 0 {
		var names []string
		for i, failure := range failures {
			if i >= 5 {
				break
			}
			names = append(names, fmt.Sprintf("%s (%s)", failure.candidate.Filename, failure.err))
		}
		r.Logger.Warningf("%d file(s) have now failed %d+ times and no longer fail the run; they retry on backoff (up to %d days) and stay visible as failed= in this summary: %s",
			len(failures), MaxFatalAttempts, int(MaxRetryBackoff.Hours()/24), strings.Join(names, ", "))
	}
	return summary, nil
}

// recordFailures persists each failed attempt and returns the ones that
// must fail the run.
//
// A failure keeps failing the run until the same file has been attempted
// MaxFatalAttempts times AND an upload has succeeded since that streak began.
// The second condition is what keeps a real outage (revoked Photos access,
// dead network) loudly red instead of quietly "green with failures": if
// nothing has succeeded since the failures started, every failure is still
// treated as fatal no matter how many attempts it has.
func (r *Runner) recordFailures(failures []failedCandidate) ([]error, error) {
	var retryable []error
	if r.State == nil {
		for _, failure := range failures {
			retryable = append(retryable, failure.err)
		}
		return retryable, nil
	}
	now := r.Now()
	latestSuccess, hasSuccess, err := r.State.LatestSuccessAt()
	if err != nil {
		return nil, err
	}
	for _, failure := range failures {
		entry, err := r.State.MarkFailure(SourceTypeAssetFile, failure.candidate.StateID(), failure.candidate.Fingerprint(), failure.err.Error(), now)
		if err != nil {
			return nil, err
		}
		firstFailure, hasFirst := common.TryParseISO(entry.FirstFailureAt)
		proven := hasSuccess && hasFirst && !latestSuccess.Before(firstFailure)
		if entry.FailureCount < MaxFatalAttempts || !proven {
			retryable = append(retryable, failure.err)
		}
	}
	return retryable, nil
}

func (r *Runner) isStateComplete(candidate Candidate) (bool, error) {
	if r.State == nil {
		return false, nil
	}
	return r.State.IsComplete(SourceTypeAssetFile, candidate.StateID(), candidate.Fingerprint())
}

// retryNotBefore is the moment a previously failed candidate may be
// attempted again; deferred is false when it may run now.
func (r *Runner) retryNotBefore(candidate Candidate) (retryAt time.Time, deferred bool, err error) {
	if r.State == nil {
		return time.Time{}, false, nil
	}
	entry, ok, err := r.State.EntryFor(SourceTypeAssetFile, candidate.StateID())
	if err != nil || !ok || entry.FailureCount <= 0 {
		return time.Time{}, false, err
	}
	if entry.Fingerprint != candidate.Fingerprint() {
		// The asset changed in Photos: a different file to upload, so the
		// old streak says nothing about it.
		return time.Time{}, false, nil
	}
	lastFailure, ok := common.TryParseISO(entry.LastFailureAt)
	if !ok {
		return time.Time{}, false, nil
	}
	retryAt = lastFailure.Add(RetryDelay(entry.FailureCount))
	if r.Now().Before(retryAt) {
		return retryAt, true, nil
	}
	return time.Time{}, false, nil
}

func (r *Runner) logDeferred(deferred []deferredCandidate) {
	earliest := deferred[0].retryAt
	var names []string
	for i, item := range deferred {
		if item.retryAt.Before(earliest) {
			earliest = item.retryAt
		}
		if i < 5 {
			names = append(names, item.candidate.Filename)
		}
	}
	r.Logger.Warningf("Deferred %d previously failed file(s) still in retry backoff (earliest retry %s): %s",
		len(deferred), common.ISOFormat(earliest), strings.Join(names, ", "))
}

func (r *Runner) uploadCandidate(index, total int, candidate Candidate, exported ExportedFile) error {
	contentSHA256, err := common.FileSHA256(exported.Path)
	if err != nil {
		return err
	}
	capturedAt := candidate.CapturedAt
	if capturedAt == "" {
		capturedAt = r.Now().UTC().Format("2006-01-02T15:04:05")
	}
	r.Logger.Infof("[%d/%d] Uploading %s (%s, %s)", index, total, exported.Filename, candidate.Role, formatBytes(exported.SizeBytes))
	stored, err := r.Client.UploadPhotoFile(exported.Path, capturedAt, exported.Extension, exported.MimeType, contentSHA256)
	if err != nil {
		return err
	}
	envelope, err := BuildPhotoMetadata(Envelope{
		Source:          PhotoSource,
		Account:         r.Account,
		NativeID:        candidate.NativeID,
		Role:            candidate.Role,
		Filename:        exported.Filename,
		MimeType:        exported.MimeType,
		SizeBytes:       exported.SizeBytes,
		ContentSHA256:   contentSHA256,
		UploadedAt:      common.ISOFormat(r.Now()),
		Width:           candidate.Width,
		Height:          candidate.Height,
		CapturedAt:      capturedAt,
		CaptureTZOffset: candidate.CaptureTZOffset,
		CameraMake:      candidate.CameraMake,
		CameraModel:     candidate.CameraModel,
		RecordKey:       "apple_record",
		Record:          candidate.AppleRecord,
	})
	if err != nil {
		return err
	}
	dedup := ProvenanceDedupSHA256(PhotoSource, r.Account, candidate.NativeID, candidate.Role, contentSHA256)
	if _, err := r.Client.UploadPhotoMetadata(envelope, capturedAt, contentSHA256, dedup); err != nil {
		return err
	}
	if r.State != nil {
		return r.State.MarkSuccess(SourceTypeAssetFile, candidate.StateID(), candidate.Fingerprint(), r.Now(), contentSHA256, stored.StorageKey)
	}
	return nil
}

// formatBytes mirrors the Python format_bytes ("123 B", "1.5 KiB", ...).
func formatBytes(count int64) string {
	return common.FormatBytes(count)
}
